<a href="https://colab.research.google.com/github/OlegV12/Stats/blob/Lesson_6/mathStats.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from math import factorial, pow, e, pi

In [34]:
import numpy as np
from scipy.stats import norm
import pandas as pd

In [ ]:

def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

In [ ]:
def bernulli(n, k, p):
  """
  :param n number of observations
  :param k number of occurrences
  :param p probability of occurrence
  """
  return combinations(n, k) * pow(p, k) * pow((1-p), (n-k))

In [ ]:
def poisson(n, k, p):
  """
  :param n number of observations
  :param k number of occurrences
  :param p probability of occurrence
  """
  a = n * p # lambda

  return pow(a, k) / factorial(k) * pow(e, (-a))



In [ ]:
def mean_sqrt_dev(a):
  res = []
  mean = sum(a) / len(a)
  for i in a:
    res.append(pow((i - mean), 2))
  return np.sqrt(sum(res) / len(a)), np.sqrt(sum(res) / (len(a)-1))


In [ ]:
def prob(mu, x, sqd):
  return (x-mu)/sqd

Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks):
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].
Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с помощью функции cov из numpy
Полученные значения должны быть равны.
Найдите коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков,
а затем с использованием функций из библиотек numpy и pandas.

In [18]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

In [8]:
zp_ks = list(a*b for a, b in zip(zp, ks))
zp_ks

[14035, 25830, 166060, 183800, 18360, 51730, 35262, 135300, 89520, 91520]

In [28]:
cov = np.mean(zp_ks) - (np.mean(zp) * np.mean(ks))
np.round(cov, 2)

9157.84

In [27]:
# проверка 
np_cov = np.cov(zp, ks, ddof=0)
np_cov[0][1]

9157.84

In [31]:
correlation = cov / (np.std(zp)*np.std(ks))
correlation

0.8874900920739158

In [33]:
# проверка через numpy
np_correlation = np.corrcoef(zp, ks)
np_correlation[0][1]

0.8874900920739162

In [37]:
# проверка через pandas
pd_zp = pd.Series(zp)
pd_ks = pd.Series(ks)
pd_zp.corr(pd_ks)

0.8874900920739162

Измерены значения IQ выборки студентов,
обучающихся в местных технических вузах:
131, 125, 115, 122, 131, 115, 107, 99, 125, 111.
Известно, что в генеральной совокупности IQ распределен нормально.
Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [39]:
iq = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])
iq_mean = np.mean(iq)

In [40]:
a1 = iq_mean - (2.262 * (np.std(iq, ddof=1)/np.sqrt(10)))
a1

110.55660776308164

In [41]:
a1 = iq_mean + (2.262 * (np.std(iq, ddof=1)/np.sqrt(10)))
a1

125.64339223691834

Ответ: доверительный инервал с надежностью 0.95  [110.56, 125.64]

Известно, что рост футболистов в сборной распределен нормально
с дисперсией генеральной совокупности, равной 25 кв.см. Объем выборки равен 27,
среднее выборочное составляет 174.2. Найдите доверительный интервал для математического
ожидания с надежностью 0.95.

In [42]:
a1 = 174.2 - (1.96 * (5 / np.sqrt(27)))
a1

172.31398912064722

In [43]:
a2 = 174.2 + (1.96 * (5 / np.sqrt(27)))
a2

176.08601087935276

Ответ: доверительный инервал с надежностью 0.95 [172.31, 176.09]